mediorum/server/cuckoo.go

```diff
@@ -0,0 +1,159 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	cuckoo "github.com/panmari/cuckoofilter"
+)
+
+var (
+	myCuckooKeyName = "my_cuckoo"
+	cuckooMap       = map[string]*cuckoo.Filter{}
+	cuckooMu        = sync.RWMutex{}
+)
+
+func (ss *MediorumServer) serveCuckooLookup(c echo.Context) error {
+	cid := c.Param("cid")
+	hosts := ss.cuckooLookup(cid)
+	return c.JSON(200, hosts)
+}
+
+func (ss *MediorumServer) serveCuckooSize(c echo.Context) error {
+	sizes := map[string]any{}
+	cuckooMu.RLock()
+	for host, filter := range cuckooMap {
+		sizes[host] = map[string]any{
+			"size":        filter.Count(),
+			"load_factor": filter.LoadFactor(),
+		}
+	}
+	cuckooMu.RUnlock()
+	return c.JSON(200, sizes)
+}
+
+func (ss *MediorumServer) serveCuckoo(c echo.Context) error {
+	ctx := c.Request().Context()
+	r, err := ss.bucket.NewReader(ctx, myCuckooKeyName, nil)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// would be nice to do some Last-Modified stuff here
+	return c.Stream(200, "", r)
+}
+
+func (ss *MediorumServer) cuckooLookup(cid string) []string {
+	hosts := []string{}
+	cidBytes := []byte(cid)
+	cuckooMu.RLock()
+	for host, filter := range cuckooMap {
+		if filter.Lookup(cidBytes) {
+			hosts = append(hosts, host)
```

Comment:
False positive:  0.01%

node1, node2, node3

1. is host up
2. does have file on disk